cmd/tsconnect/wasm/wasm_js.go

```diff
@@ -324,6 +349,117 @@ func (i *jsIPN) logout() {
 	go i.lb.Logout()
 }
 
+func (i *jsIPN) tcp(host string, port int, readCallback js.Value, connectTimeoutSeconds int) any {
+	portInt := strconv.Itoa(port)
+	addr := net.JoinHostPort(host, portInt)
+
+	return makePromise(func() (any, error) {
+		if readCallback.Type() != js.TypeFunction {
+			return nil, errors.New("Invalid readCallback received")
+		}
+
+		connectTimeout := time.Duration(connectTimeoutSeconds) * time.Second
+		jsTCPSession, err := i.NewJSTCPSession(addr, readCallback, connectTimeout)
+		if err != nil {
+			return nil, err
+		}
+
+		jsObjectWrapper := jsTCPSession.jsWrapper()
+		return jsObjectWrapper, nil
+	})
+
+}
+
+type jsTCPSession struct {
+	jsIPN       *jsIPN
+	conn        net.Conn
+	writeBuffer []byte
+	readBuffer  []byte
+}
+
+const defaultTCPConnectTimeout time.Duration = 5 * time.Second
+
+func (i *jsIPN) NewJSTCPSession(addr string, readCallback js.Value, connectTimeout time.Duration) (*jsTCPSession, error) {
+	if connectTimeout == 0 {
+		connectTimeout = defaultTCPConnectTimeout
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), connectTimeout)
+	defer cancel()
+
+	conn, err := i.dialer.UserDial(ctx, "tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &jsTCPSession{
+		jsIPN:       i,
+		conn:        conn,
+		writeBuffer: make([]byte, 1024),
+		readBuffer:  make([]byte, 4*1024*1024),
```

Comment:
maybe the size should be configurable same for the writeBuffer

Reply:
that's quite the read buffer.  BTW, another Go idiomatic way to write 4 MB would be "4 << 20". But I think 1<<20 is probably enough?

and why are write and read so different?

Reply:
Honestly at first i just chose a kb for each but for vnc a big buffer just worked way better. [novnc even uses 4 mb internally](https://github.com/novnc/noVNC/blob/cbbd9ab069e384d230c032f51bf4e713ed02766e/core/websock.js#L56) so I picked that for the moment. I think this should be configurable with the defaults you specified as for the VNC usecase 4 megs just work way better.

To answer your second question the behaviour is different because this way the writer can implicitly choose the buffer size using the size property on the JS ArrayBuffer. Thinking about the read side tho, in hindsight my code does not make alot of sense. So I will simplify that part and just use the size of the go buffer (which is the case anyhow atm just with some useless code).